ONIX Source v0.3.0 updates (#104)
open-ephys/gui-docs: source/User-Manual/Plugins/Onix-Source.rst

JSON file to load. The loaded JSON file will be used to configure the probe settings, such as the
selected electrodes. 

Automatically saved ProbeInterface JSON Files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. note:: Requires >= v0.3.0 of the plugin, and >= v1.0.2 of the GUI.

Every time a recording is started all data streams that contain ProbeInterface

> metadata will save the ProbeInterface data in a JSON file. The files will be saved in the folder

Reviewer: ```suggestion
Every data stream that contains ProbeInterface metadata will write that metadata to a JSON file every time a recording is started. The files will be saved in the folder
```

> path set by the recording options interface at the top of the GUI's main window; note that this

Reviewer: ```suggestion
path set by the recording options interface at the top of the GUI's main window. This
```

   "**Probe**", "**Channels**", "**Plugin Version**"
   "Neuropixels 1.0", "384 AP, 384 LFP", "≥0.1.0"
   "Neuropixels 2.0 (quad-shank)", "384 wideband", "≥0.1.0"
   "Neuropixels 2.0 (single-shank)", "384 wideband", "≥0.3.0"

.. note:: 
  Prior to v0.3.0, the plugin could not detect which probe was connected, or if it was compatible.

>   Beginning in v0.3.0, if a probe is not compatible a pop-up window will explain that it cannot be configured.

Reviewer: When talking about version differences, I tend to omit what was left out in a previous version and only mention what has been added in the new current version.

  The quaternion data can be used to drive the commutator in the GUI, allowing for real-time
  commutation of the headstage. See :ref:`commutatorcontrol` for more information on how to use the commutator with the BNO055 IMU data.

Loading Data
============

There are two major platforms that can be used to load and process data recorded by the Open Ephys
GUI and the ONIX Source plugin. First is the `open-ephys-python-tools
<https://pypi.org/project/open-ephys-python-tools/>`__ which is maintained by the Open Ephys GUI
team. The second is `SpikeInterface <https://pypi.org/project/spikeinterface/>`__, which is more

> focused on Neuropixels recording and analysis.

Reviewer: ```suggestion
team. The second is `SpikeInterface <https://pypi.org/project/spikeinterface/>`__, which is oriented towards Neuropixels recording/analysis but can be used for ephys from any source.
```

maybe?

Author: Good point, I like this wording.

open-ephys-python-tools
#########################

The Python tools can be installed by running the following line inside a Python virtual environment:

.. code-block::

  pip install open-ephys-python-tools

The following script shows how to discover what was recorded and inspect all data streams in an

> ONIX Source session:

Reviewer: ```suggestion
The following script prints information about all the data streams recorded by the first record node
in your signal chain during a single acquisition session:
```

I don't see this script as anything ONIX-specific. Also my interpretation of how the script works according to your comments is that only the first record node is inspected, so maybe this description is a little more correct.

  # Point to the top-level data directory (the parent folder containing the Record Node folder(s))
  session = Session('./path/to/data')

  if session.recordnodes is None:
    raise ValueError("No valid Record Nodes found.")

  # List all record nodes
  for node in session.recordnodes:
    print(node)

  # Iterate through each experiment (recording) in the first record node
  node = session.recordnodes[0]
  for recording in node.recordings:
    print(f"\nExperiment {recording.experiment_index}")

    if recording.continuous is None:
      print("No continuous channels found.")
      continue

    for stream in recording.continuous:
        print(f"  Stream: {stream.metadata.stream_name}")
        print(f"    Sample rate: {stream.metadata.sample_rate} Hz")
        print(f"    Channels: {stream.samples.shape[1]}")

To load samples from a stream, call :code:`samples[]` on the stream variable. Similarly, the
timestamps can be loaded by calling :code:`timestamps[]` on the stream.

.. code-block:: python

  from open_ephys.analysis.session import Session

  session = Session('./path/to/data')
  node = session.recordnodes[0]
  recording = node.recordings[0]

>   stream = recording.continuous[0] # "BreakoutBoard-DigitalIO" for the DigitalIO stream

Reviewer: `"BreakoutBoard-DigitalIO" for the DigitalIO stream for the DigitalIO stream`

I'm not sure this is will necessarily be true for the recording that the user might be trying load. Maybe I would change the comment to say that there must be at least one continuous recording to be able to load a recording located at index 0

  samples = stream.samples[:,0]  # Get all samples from the first channel
  timestamps = stream.timestamps # Get all timestamps

The following script shows how to extract event markers from event channels and overlay them on top
of continuous traces. In this example, the DigitalIO stream is used for both the event and the
continuous channel, but any time-synchronized channel recorded by the ONIX Source plugin will have

> the same timestamps and can be overlaid.

Reviewer: ```suggestion
of continuous traces. In this example, the DigitalIO continuous and event streams are assumed to be recorded. Any time-synchronized channel recorded by the ONIX Source plugin will have
the same timestamps and can be overlaid.
```

"Any time-synchronized channel recorded by the ONIX Source plugin will have
the same timestamps and can be overlaid." maybe this part should go above?

.. code-block:: python

  from open_ephys.analysis.session import Session
  import matplotlib.pyplot as plt

  session = Session('/path/to/data')
  node = session.recordnodes[0]
  recording = node.recordings[0]
  digital_io = recording.continuous["BreakoutBoard-DigitalIO"]

  plt.figure()

  stream_name = 'BreakoutBoard-DigitalIO'

  plt.plot(digital_io.timestamps, digital_io.samples[:,7])

  plt.title(f"{digital_io.name.split('.')[1][:-1]}")

  for i, event in recording.events.iterrows(): # type: ignore
      if event['stream_name'] == stream_name:
          if event['state'] == 1:
              start = event['timestamp']
              plt.axvline(x=start, color='r')

          elif event['state'] == 0:
              end = event['timestamp']
              plt.axvline(x=end, color='g')
          
  plt.show()

SpikeInterface
#########################

> 

Reviewer: Might be worth (re-?)introducing SpikeInterface here, for example:

SpikeInterface is a Python package specialized for loading and analyzing Neuropixels data.

Integration with SpikeInterface and ProbeInterface for the ONIX Source plugin has been added as of
version v0.103.1 (SpikeInterface) and v0.3.1 (ProbeInterface). Note that this integration requires
version 0.3.0 of the ONIX Source plugin to correctly integrate the data; all recordings saved prior

> to v0.3.0 will not correctly load certain metadata, such as the Neuropixels sample shifts.

Reviewer: ```suggestion
SpikeInterface v0.103.1+ and ProbeInterface V0.3.1+ support reading data from the ONIX Source plugin. 

.. note:: ONIX Source plugin v0.3.0+ is required to correctly integrate Neuropixels metadata; recordings saved prior to v0.3.0 will not correctly load certain metadata, such as the Neuropixels sample shifts.
```

SpikeInterface can be installed by running the following line inside a Python virtual environment:

.. code-block::

  pip install spikeinterface

The following script shows how to list all available streams in a Record Node folder.

.. code-block:: python

  import spikeinterface.extractors as se

  folder_path = '/path/to/data'

  # List all available streams without loading any data
  stream_names, stream_ids = se.OpenEphysBinaryRecordingExtractor.get_streams(folder_path)
  print("Available streams:")
  for i, name in enumerate(stream_names):
      print(f"  {i}: {name}")

Once the stream ID or stream name is known, the following script can be used to load the chosen
stream. Probe metadata from the `automatically-saved ProbeInterface JSON files
<#automatically-saved-probeinterface-json-files>`__ is loaded automatically when using a recording
made with plugin version ≥ v0.3.0:

> .. code-block:: python

Reviewer: does this script assume the recording node directory and the gui saving directory are the same? if so, I would mention that.

  import spikeinterface.extractors as se

  folder_path = '/path/to/data'

  # Load a specific stream by ID (replace with your actual stream ID from the list above)
  recording = se.read_openephys(folder_path, stream_id='2') 

  print(f"\nChannels: {recording.get_num_channels()}")
  print(f"Sampling rate: {recording.get_sampling_frequency()} Hz")
  print(f"Duration: {recording.get_total_duration():.2f} s")

For recordings with multiple experiments, use the :code:`block_index` parameter to select which
experiment to load:

.. code-block:: python

  import spikeinterface.extractors as se

  folder_path = '/path/to/data'

  # Determine how many experiments were recorded
  n_blocks = se.OpenEphysBinaryRecordingExtractor.get_num_blocks(folder_path)
  print(f"Number of experiments: {n_blocks}")

  for block_index in range(n_blocks):
      recording = se.read_openephys(
          folder_path,

>           stream_id='2',

Reviewer: what is `stream_id='2'`?

Author: Good point, I didn't make any assumptions in this script; I tried to set it up so that the user would use the previous script to discover the streams, and then choose their own ID in this one, but using '2' is confusing. I changed it to '0', in case they only have one valid stream

          block_index=block_index
      )
      print(f"Experiment {block_index}: {recording.get_total_duration():.2f} s")

Additional Plugin Integration
===============================

Neuropixels-CAR
##################

Neuropixels probe streams are automatically detected by the :ref:`neuropixelscar` plugin, which uses
a Common Average Reference across the specific groups of channels belonging to the same ADC.

> .. note:: First introduced in v0.3.0.

Reviewer: ```suggestion
.. note:: Requires Onix Source plugin v0.3.0+.
```

LFP Viewer
##################

Neuropixels probe streams in the :ref:`lfpviewer` can be sorted by depth, according to the electrode
metadata defined by the ProbeInterface specification above. Additionally, the channels can be
colored by shank, to visually indicate when a channel belongs to a different shank than its
neighbor.

> .. note:: First introduced in v0.3.0.

Reviewer: ```suggestion
.. note:: Requires Onix Source plugin v0.3.0+.
```
